File: fabric/src/main/java/net/shiroha233/roadweaver/client/fabric/OpenPresetEditorEntry.java
package net.shiroha233.roadweaver.client.fabric;

import me.shedaniel.clothconfig2.api.AbstractConfigListEntry;
import me.shedaniel.math.Rectangle;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.events.GuiEventListener;
import net.minecraft.client.gui.narration.NarratableEntry;
import net.minecraft.client.input.MouseButtonEvent;
import net.minecraft.network.chat.Component;
import net.shiroha233.roadweaver.client.config.MaterialPresetEditorScreen;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * 材料预设编辑器入口
 */
public class OpenPresetEditorEntry extends AbstractConfigListEntry<Void> {
    private Rectangle lastArea;

    public OpenPresetEditorEntry() {
        super(Component.translatable("config.roadweaver.open_preset_editor"), false);
    }

    @Override
    public Void getValue() {
        return null;
    }

    public void setValue(Void value) {}

    @Override
    public Optional<Void> getDefaultValue() {
        return Optional.empty();
    }

    @Override
    public void render(GuiGraphics g, int index, int y, int x, int entryWidth, int entryHeight,
                       int mouseX, int mouseY, boolean isHovered, float delta) {
        super.render(g, index, y, x, entryWidth, entryHeight, mouseX, mouseY, isHovered, delta);
        this.lastArea = getEntryArea(x, y, entryWidth, entryHeight);

        Font font = Minecraft.getInstance().font;
        Component label = getDisplayedFieldName();
        int color = getPreferredTextColor();
        int textY = y + (entryHeight - font.lineHeight) / 2;
        g.drawString(font, label, x + 4, textY, color, false);
    }

    @Override
    public boolean mouseClicked(MouseButtonEvent event, boolean doubleClick) {
        if (event.button() == 0 && lastArea != null && lastArea.contains(event.x(), event.y())) {
            Minecraft mc = Minecraft.getInstance();
            if (mc != null) {
                mc.setScreen(new MaterialPresetEditorScreen(mc.screen));
            }
            return true;
        }
        return super.mouseClicked(event, doubleClick);
    }

    @Override
    public List<? extends GuiEventListener> children() {
        return Collections.emptyList();
    }

    @Override
    public List<? extends NarratableEntry> narratables() {
        return Collections.emptyList();
    }
}

File: fabric/src/main/kotlin/net/shiroha233/roadweaver/client/fabric/OpenPresetEditorEntry.kt
package net.shiroha233.roadweaver.client.fabric

import me.shedaniel.clothconfig2.api.AbstractConfigListEntry
import me.shedaniel.math.Rectangle
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.gui.components.events.GuiEventListener
import net.minecraft.client.gui.narration.NarratableEntry
import net.minecraft.network.chat.Component
import net.shiroha233.roadweaver.client.config.RoadPresetEditorScreen
import java.util.Optional

class OpenPresetEditorEntry : AbstractConfigListEntry<Void>(
    Component.translatable("config.roadweaver.open_preset_editor"),
    false
) {
    private var lastArea: Rectangle? = null

    private class DummyParentScreen : Screen(Component.empty())

    override fun getValue(): Void? = null

    override fun getDefaultValue(): Optional<Void> = Optional.empty()

    override fun render(
        g: GuiGraphics,
        index: Int,
        y: Int,
        x: Int,
        entryWidth: Int,
        entryHeight: Int,
        mouseX: Int,
        mouseY: Int,
        isHovered: Boolean,
        delta: Float
    ) {
        super.render(g, index, y, x, entryWidth, entryHeight, mouseX, mouseY, isHovered, delta)
        lastArea = getEntryArea(x, y, entryWidth, entryHeight)

        val font: Font = Minecraft.getInstance().font
        val label: Component = displayedFieldName
        val color: Int = preferredTextColor
        val textY = y + (entryHeight - font.lineHeight) / 2
        g.drawString(font, label, x + 4, textY, color, false)
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        val area = lastArea
        if (button == 0 && area != null && area.contains(mouseX, mouseY)) {
            val mc = Minecraft.getInstance()
            mc.setScreen(RoadPresetEditorScreen(mc.screen ?: DummyParentScreen()))
            return true
        }
        return false
    }

    override fun children(): List<GuiEventListener> = emptyList()

    override fun narratables(): List<NarratableEntry> = emptyList()
}
